### PR TITLE
feat(SF-1022): Add `prevReq` param to overrides function

### PR DESCRIPTION
**Pull request by Vic-Dev**
* Call overrides functions with currReq and prevReq
* Remove search overrides call in autocompleteProducts

### CODE REVIEW

**Comment by Vic-Dev** [comment]
TODO: don't forget to update core V!!!

File: src/core/requests.ts [hunk 1]
```diff
@@ -8,10 +8,27 @@ import SearchAdapter, { MAX_RECORDS } from './adapters/search';
...
 namespace Requests {
+  export interface PastRequests {
+    search: Partial<Request>;
+    autocompleteProducts: Partial<QueryTimeProductSearchConfig>;
+  }
 
-  export const search = (state: Store.State): Request => {
+  export const pastReqs: Requests.PastRequests = {
+    search: {},
+    autocompleteProducts: {}
```
Comment (jmbtutor) [inline src/core/requests.ts]: ... and if it's whole, these should be cast.

File: src/core/utils.ts [hunk 2]
```diff
@@ -39,3 +39,10 @@ export const sortBasedOn = function<T,S>(toBeSorted: T[], basisArray: S[], callb
   });
   return output.concat(ids);
 };
+
+// tslint:disable-next-line max-line-length
+export function normalizeToFunction<T>(objOrFn: Partial<T> | ((...arg: Partial<T>[]) => Partial<T>)): (arg: Partial<T>) => Partial<T> {
```
Comment (jmbtutor) [inline src/core/utils.ts]: Should just be `T` inside the function type.

File: test/unit/core/utils.ts [hunk 3]
```diff
@@ -44,4 +44,19 @@ suite('utils', ({ expect, spy, stub }) => {
       ]);
     });
   });
+
+  describe('normalizeToFunction()', () => {
+    it('should return the given argument if its a function', () => {
+      const fn = (r) => r;
+
+      expect(utils.normalizeToFunction(fn)).to.eql(fn);
```
Comment (jmbtutor) [inline test/unit/core/utils.ts]: `eq`

File: src/core/configuration.ts [hunk 4]
```diff
@@ -199,19 +202,27 @@ namespace Configuration {
...
-      products?: QueryTimeProductSearchConfig | ((config: QueryTimeProductSearchConfig) => QueryTimeProductSearchConfig);
+      suggestions?: Configuration.autocompleteSuggestionsOverrides;
+      // tslint:disable-next-line max-line-length
+      products?: Configuration.AutocompleteProductsOverrides;
     };
   }
 
+  // tslint:disable:max-line-length
+  export type autocompleteSuggestionsDefaults = QueryTimeAutocompleteConfig | ((config: QueryTimeAutocompleteConfig) => QueryTimeAutocompleteConfig);
+  export type autocompleteSuggestionsOverrides = QueryTimeAutocompleteConfig | ((currConfig: QueryTimeAutocompleteConfig, prevConfig: QueryTimeAutocompleteConfig) => QueryTimeAutocompleteConfig);
```
Comment (jmbtutor) [inline src/core/configuration.ts]: Types should start with a capital letter.

File: src/core/requests.ts [hunk 5]
```diff
@@ -59,18 +87,24 @@ namespace Requests {
...
-      numSearchTerms: Configuration.extractAutocompleteSuggestionCount(config),
-      numNavigations: Configuration.extractAutocompleteNavigationCount(config),
-      sortAlphabetically: Configuration.isAutocompleteAlphabeticallySorted(config),
-      fuzzyMatch: Configuration.isAutocompleteMatchingFuzzily(config)
-    }, config.autocomplete.overrides.suggestions);
-
-  export const autocompleteProducts = (state: Store.State): QueryTimeProductSearchConfig => {
+    Requests.chain(
+      Configuration.autocompleteSuggestionsDefaults(config),
+      normalizeToFunction({
```
Comment (jmbtutor) [inline src/core/requests.ts]: Move this object out and name it for readability and consistency with the other functions.

File: src/core/utils.ts [hunk 6]
```diff
@@ -39,3 +39,10 @@ export const sortBasedOn = function<T,S>(toBeSorted: T[], basisArray: S[], callb
   });
   return output.concat(ids);
 };
+
+// tslint:disable-next-line max-line-length
+export function normalizeToFunction<T>(objOrFn: Partial<T> | ((...arg: T[]) => T)): (...arg: T[]) => T {
+  return typeof objOrFn === 'function'
+    ? objOrFn
+    : (obj) => (Object.assign(obj, objOrFn));
+}
```
Comment (jmbtutor) [inline src/core/utils.ts]: Your editor should be able to add a trailing newline.

File: test/unit/core/requests.ts [hunk 7]
```diff
@@ -1,14 +1,39 @@
...
 import SearchAdapter, { MAX_RECORDS } from '../../../src/core/adapters/search';
 import Requests from '../../../src/core/requests';
 import Selectors from '../../../src/core/selectors';
+import * as utils from '../../../src/core/utils';
 import suite from '../_suite';
 
-suite('requests', ({ expect, stub }) => {
+suite('requests', ({ expect, stub, spy }) => {
+  describe('override', () => {
+    it('should return a function that calls override function with given request and pastRequest key', () => {
```
Comment (jmbtutor) [inline test/unit/core/requests.ts]: Consider "... that calls the given override function..." to make it flow better.

File: test/unit/core/requests.ts [hunk 8]
```diff
@@ -173,112 +211,101 @@ suite('requests', ({ expect, stub }) => {
...
-      };
-      const chained = { e: 'f' };
-      const biasReq = { c: 'd' };
-      const state: any = {};
-      const chain = stub(Requests, 'chain').returns(chained);
-      const search = stub(Requests, 'search').returns({ i: 'j' });
-      const realTimeBiasing = stub(Requests, 'realTimeBiasing').returns(biasReq);
-      stub(Selectors,'config').returns(config);
+      config.personalization.realTimeBiasing.autocomplete = true;
+      const realTimeBiasBuiltReq = { a: 'b' };
```
Comment (jmbtutor) [inline test/unit/core/requests.ts]: For style, consider putting this before the assignment above.

**Pull request merged**